Add Auto Deployments with CloudRun + GitHub Actions

SudharakaP opened the pull request:
This basically takes away most of the developer effort in deploying jhipster-online, and automates the deployment using CloudRun + GitHub Actions.

Fixes https://github.com/jhipster/jhipster-online/issues/172

pascalgrimaud commented:
@SudharakaP : to understand well, what happened once this PR is merged ? Will it replace definitively the original deployment in GCE ? Should we stop GCE before ?

SudharakaP commented:
@pascalgrimaud : Nothing will happen and it will not stop the GCE deployment since the GCE deployment is all manual, this PR has no effect on it. 

After we merge this PR, and create a release (with a tag say v2.6.2), then the deployment to CloudRun will trigger. Even then nothing will happen to the GCE deployment. The CloudRun deployment will be separate from the GCE deployment. 

We can then test if the CloudRun deployment is working perfectly and if so, we have to setup the custom domain in CloudRun (so that CloudRun points to https://start.jhipster.tech/). After that we can manually shutdown the GCE. 

Bottom line this PR merging will not automatically do anything to GCE. We have to manually shut it down after we do the testing and we are totally sure the CloudRun deployment works as expected. :smile:

pascalgrimaud commented:
@SudharakaP : nice ! I though about https://start-beta.jhipster.tech/ or something else

PierreBesson commented:
I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.

Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

vishal423 commented on Dockerfile:
```diff
@@ -7,7 +7,7 @@ RUN \
     chmod +x /code/mvnw && \
     sleep 1 && \
-    ./mvnw package -Pprod -DskipTests && \
+    ./mvnw package -Pprod -Pgcp -DskipTests && \
```
If we are using this docker file for exclusive gcp deployments, then, I suggest renaming to reflect the context. Further, the maven `gcp` profile includes the spring `prod` profile.

SudharakaP replied:
You are correct; I've changed it to only include the gcp profile. As for renaming the docker file; currently that's not possible. Because, `gcloud builds submit` command expects the file to be named as `Dockerfile`. :cry:

SudharakaP commented:
> I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.
> 
> Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

@PierreBesson : Note that the deployment action is only triggered when creating a tag and therefore I believe only people in the core team can do this. Just by pushing things to branches won't trigger the deployment action. :smile:

vishal423 commented on src/main/resources/config/application-prod.yml:
```diff
@@ -26,6 +26,12 @@ management:
         enabled: false
 
 spring:
+  cloud:
```
I suggest moving these to separate `gcp` specific profile

SudharakaP replied:
Thanks. Done. :+1:

vishal423 commented on src/main/resources/config/application-prod.yml:
```diff
@@ -44,16 +50,21 @@ spring:
         prepStmtCacheSize: 250
         prepStmtCacheSqlLimit: 2048
         useServerPrepStmts: true
+        socketFactory: com.google.cloud.sql.mysql.SocketFactory
```
same as above, production profile properties shouldn't be tied with `gcp`

SudharakaP replied:
Thanks. Done. :+1:

vishal423 commented on src/main/resources/config/application-prod.yml:
```diff
@@ -44,16 +50,21 @@ spring:
     contexts: prod
   mail:
-    host: localhost
+    host: $SECRET_VALUE
```
any specific reason to use `$SECRET_VALUE` in all places? I prefer to revert unless they serve a specific purpose

SudharakaP replied:
Good point. I thought about it for a moment and put it there just so that we could keep track of which values are being replaced. However yes it can be removed. :smile:

SudharakaP replied:
Done. :+1:

vishal423 commented on src/main/resources/config/application-prod.yml:
```diff
@@ -78,7 +89,7 @@ spring:
 server:
   use-forward-headers: true
-  port: 8080
+  port: ${PORT}
```
you can always override with environment variables like `SERVER_PORT`

SudharakaP replied:
This PORT variable is GCP specific. It has nothing to do with a GiHub environment variable. :smile:

SudharakaP commented:
> I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.
> 
> Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

@PierreBesson : I've changed the deployment action to be only triggered when publishing a release as I think it's even better than relying on the tag. That means only people in the core team or people who has access to create releases can make the internal decryption happen. Even then they will not see the secrets since they are not recorded anywhere. :smile:

vishal423 commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,88 @@
+on:
+  push:
+    tags:
+      - 'v*.*.*'
```
It would be good to restrict this workflow to the `master` branch

SudharakaP replied:
Thanks. Just did it a while ago after Pierre's comment above. I've restricted for the workflow to be triggered when publishing releases; which I think is better. :smile:

vishal423 commented on src/main/resources/config/application-gcp.yml:
```diff
@@ -0,0 +1,50 @@
+      password:
+server:
+  port: ${PORT}
+jhipster:
+  security:
+    authentication:
+      jwt:
+        base64-secret:
+  mail: # specific JHipster mail property, for standard properties see MailProperties
+    enabled: true
+    from: jhipster-online@jhipster.tech
+    base-url: https://start.jhipster.tech
+application:
+  github:
+    client-id:
+    client-secret:
+    jhipster-bot-oauth-token:
+  gitlab:
+    client-id:
+    client-secret:
+    redirect-uri:
```
unless these override a value provided in base application.yml or other profile-specific yml (prod/dev), I will prefer to not duplicate here.

SudharakaP replied:
Thanks :+1:

vishal423 commented on src/main/resources/config/application-gcp.yml:
```diff
@@ -0,0 +1,50 @@
+      data-source-properties:
+        socketFactory: com.google.cloud.sql.mysql.SocketFactory
+        cloudSqlInstance:
+      jdbcUrl:
+      username:
+      password:
```
unless these override a value provided in base application.yml or other profile-specific yml (prod/dev), I will prefer to not duplicate here.

SudharakaP replied:
Thanks. I've removed the duplicate properties that are present in `application-gcp.yml`. However properties such as `jdbcUrl` are not defined in `application-prod.yml` or `application.yml` so I think I have to keep them there for the substitutions. :smile: :thinking:

SudharakaP commented:
@jdubois @pascalgrimaud : Just pinging to see if we could merge this; or if anything else needs to be done? This I think would make the deployments a lot simpler. :smile:

pascalgrimaud commented:
@SudharakaP : I think you know better than me, what is the best here. But if we decide to use this new deployment, it's important to document it somewhere, in case you're in vacations for example. So we'll know how to fix when there is a problem
- how to deploy
- how to restart
- how to see log
- etc...
cc @PierreBesson too

SudharakaP commented:
> @SudharakaP : I think you know better than me, what is the best here. But if we decide to use this new deployment, it's important to document it somewhere, in case you're in vacations for example. So we'll know how to fix when there is a problem
> 
>     * how to deploy
> 
>     * how to restart
> 
>     * how to see log
> 
>     * etc...
>       cc @PierreBesson too

@pascalgrimaud : I've just added some details about these things to the readme. If you have any other things that need to be added let me know. :smile: 

Also do we do a docker hub deployment in parallel when we do a new release? Is so I think I can add some code to automate that as well. :smile:

SudharakaP commented:
@jdubois @PierreBesson : I've added the environment variables to the `app.yaml` file added a `.dockerignore` file so that the `app.yaml` file is ignored in the docker image. I hope this solves the docker image containing secrets problem. Also I've fixed the email problem we saw. Give it a try and let me know if anything else needs to be done. 😄

PierreBesson commented:
@SudharakaP It looks good to me. Let's start the migration !

SudharakaP commented:
@PierreBesson : Okay, so shall I merge this PR and try to add the custom domain. Or maybe we can do it together over a call?

PierreBesson commented:
I trust you to do it. If you need anything just ping me.

saturnism commented:
Congrats! Glad to see the migration was successful :D

SudharakaP commented:
@saturnism : Thanks much for your help on this. Greatly appreciated. 👍🏼
